MultiOnRamp - family-agnostic messages

elatoskinas opened the pull request:
## Motivation
* Converts the `EVM2EVMMultiOnRamp` to an `EVM2Any` family-agnostic ramp to support non-EVM use-cases from a single multi-onramp.
* Collapses `sourceTokenData` + `tokenAmounts` into one `RampTokenAmount` array (simplification to no longer require encoding / decoding on-chain)

## Solution
* Message format changes: convert `EVM2EVMMessage` to `EVM2AnyRampMessage`, which contains a more generic `receiver` target and stores the family-specific args as bytes
  * Note: `hash(Any2EVMRampMessage) != hash(EVM2AnyRampMessage)` - the messages are now asymmetric
  * Note: `EVM2AnyRampMessage.messageId = hash(EVM2AnyRampMessage)`, but `messageId != hash(Any2EVMRampMessage)` due to the message asymmetry
* Add a `familyTag` to the dest chain config, to group multiple chains under one family to perform validations
* Convert `extraArgs` to bytes, implement conditional flows to retrieve EVM-specific fields
* Add `RampTokenAmount` which is constructed from `Client.EVMTokenAmount` on the OnRamp. The struct unrolls `SourceTokenData` and `EVMTokenAmount` into one struct, no longer requiring additional decoding
* Out-of-scope: moving family-specific fee & validation logic to `PriceRegistry` - to be done in next PR

elatoskinas committed:
feat: convert multi-onramp to be family agnostic

elatoskinas commented on contracts/src/v0.8/ccip/libraries/Client.sol:
```diff
@@ -52,4 +52,33 @@ library Client {
+  /// @param destChainExtraArgs: extra args to send to the dest chain
+  struct ExtraArgsV1 {
+    bool allowOutOfOrderExecution;
+    bytes4 destChainArgsVersion;
```
note: I've explored a few alternatives for the args

# Current Solution
* `extraArgs` format does not change, other than being converted to `bytes`. We still use `EVMExtraArgsV1` and `EVMExtraArgsV2`
* `EVM2AnyRampMessage` encodes `extraArgs` with the **lates family-specific version**. If the user passes in `EVMExtraArgsV1` or no extra args, the resulting `extraArgs` will always be `EVMExtraArgsV2`. This will also simplify off-chain handling
* Since `extraArgs` are bytes, they are conditionally decoded in flows where EVM-specific fields are required, such as `getFee` calculations and `allowOutOfOrderExecution`
* The main drawback of this approach compared to previous solutions is that we do not utilize shared fields between families (such as allowing out of order execution, or other potential fields in the future). However, the solution is much more simpler and inline with the previous ExtraArgs interfaces, so it will be much easier to adapt to for the users as they do not have to switch to a new extra args format.

# Previous solutions

## Option 1 - selector-based versioning

```solidity
abi.encodeWithSelector(
  EVM_FAMILY_TAG
  abi.encodeWithSelector(
    V1_TAG,
    abi.encode(destChainArgsStruct)
  )
)
```

The main issues with this approach are:
* No way to slice `bytes memory` arrays - so removing the selector from the bytes becomes more complicated (see: [SO thread](https://stackoverflow.com/questions/74443594/how-to-slice-bytes-memory-in-solidity)). `abi.decodeWithSelector` is not (yet) available (though it [will be included eventually in Solidity](https://github.com/ethereum/solidity/issues/14996))
* Because of the above 2 limitations, the complexity of decoding dest-chain specific args for validations would be much higher
* Note on slicing:
  * We could slice calldata `message.extraArgs`, however this would be complex because we need legacy EVMExtraArgs support. Sometimes we need a slice of `[4:` (for EVMExtraArgsV1, EVMExtraArgsV2), and in the new args case, we need a slice of `[8:` (to skip the `FAMILY_TAG`, and the `VERSION_TAG`)

## Option 2 - family + args selector

```solidity
// separate tag for family args
bytes4 public constant EVM_FAMILY_ARGS_V1 = bytes4(keccak256("CCIP EVM Family Args V1");
```

The main issue with this approach are:
* Validations become more complicated - when we validate the `EVM_FAMILY_TAG`, we need to additionally validate that the `destChainExtraArgs` selector pertains to the EVM family (adding a duplicated validation, i.e. `familyTag == EVM_FAMILY_TAG && (argsSelector == EVM_FAMILY_ARGS_V1 || argsSelector == EVM_FAMILY_ARGS_V2 || ...)`
* Same issue as in Option 1 - we cannot slice `bytes memory` to retrieve the selector easily

## Option 3
Approach that was replaced in: https://github.com/smartcontractkit/ccip/pull/1140/commits/2d9eb1fd50f6decb9e7818b0a57447aa304d2099

This option assumes that the `ExtraArgsV1` matches the `familyTag` (e.g. `EVM_FAMILY_TAG`), and uses the `destChainArgsVersion` to disambiguate on which `<X>FamilyExtraArgs<version>` to decode. The full `ExtraArgsV1` data is passed on to the `CCIPSendRequested` event, for the off-chain components to be able to distinguish the correct `FamilyExtraArgs` version

elatoskinas committed:
feat: merge token amount array params for multi-offramp

elatoskinas committed:
feat: merge token amount array params for multi-onramp

elatoskinas commented on contracts/src/v0.8/ccip/onRamp/EVM2EVMMultiOnRamp.sol:
```diff
@@ -28,6 +28,7 @@ contract EVM2EVMMultiOnRamp is IEVM2AnyOnRampClient, ITypeAndVersion, OwnerIsCre
 
   error CannotSendZeroTokens();
   error InvalidExtraArgsTag();
```
TODO: test coverage for new changes

RayXpub commented on contracts/src/v0.8/ccip/offRamp/EVM2EVMMultiOffRamp.sol:
```diff
@@ -783,25 +776,22 @@ contract EVM2EVMMultiOffRamp is ITypeAndVersion, MultiOCR3Base {
   /// @return destTokenAmount local token address with amount
   function _releaseOrMintSingleToken(
-    uint256 sourceAmount,
+    Internal.RampTokenAmount memory sourceTokenAmount,
```
nit: I think this can be calldata instead of memory

elatoskinas replied:
Very good find, all the `executeSingleMessage` params can flow as calldata - the gas savings are quite significant

## Using `memory`
|  |                 |         |         |         |         |
|---------------------------------------------------------------------------------------------|-----------------|---------|---------|---------|---------|
| Function Name                                                                               | min             | avg     | median  | max     | # calls |
| batchExecute                                                                                | 23565           | 202397  | 158525  | 454760  | 14      |
| execute                                                                                     | 24332           | 240967  | 154654  | 1305621 | 11      |
| executeSingleMessage                                                                        | 27496           | 100243  | 59174   | 251361  | 10      |
| executeSingleReport                                                                         | 37179           | 196051  | 113668  | 1337618 | 36      |
| manuallyExecute                                                                             | 33531           | 90090   | 50759   | 357555  | 20      |
| releaseOrMintSingleToken                                                                    | 31142           | 71823   | 69024   | 118102  | 4       |
| releaseOrMintTokens                                                                         | 30244           | 79780   | 57464   | 164792  | 8       |

## Using `calldata`
|  |                 |         |         |         |         |
|---------------------------------------------------------------------------------------------|-----------------|---------|---------|---------|---------|
| Function Name                                                                               | min             | avg     | median  | max     | # calls |
| batchExecute                                                                                | 23565           | 200376  | 157345  | 452018  | 14      |
| execute                                                                                     | 24332           | 236534  | 153496  | 1270881 | 11      |
| executeSingleMessage                                                                        | 25073           | 98611   | 58057   | 249974  | 10      |
| executeSingleReport                                                                         | 37179           | 194829  | 112510  | 1329868 | 36      |
| manuallyExecute                                                                             | 33489           | 89346   | 50633   | 356603  | 20      |
| releaseOrMintSingleToken                                                                    | 29712           | 70905   | 68107   | 117694  | 4       |
| releaseOrMintTokens                                                                         | 28220           | 77235   | 54143   | 163096  | 8       |

elatoskinas replied:
It does increase contract size by 0.8KB though 😮

elatoskinas replied:
Punting this to a new ticket - we should revisit the gas vs contract size trade-off

RayXpub commented on contracts/src/v0.8/ccip/onRamp/EVM2EVMMultiOnRamp.sol:
```diff
@@ -413,25 +454,45 @@ contract EVM2EVMMultiOnRamp is IEVM2AnyOnRampClient, ITypeAndVersion, OwnerIsCre
+    if (destChainDynamicConfig.familyTag == Client.EVM_FAMILY_TAG) {
+      Client.EVMFamilyExtraArgsV1 memory evmExtraArgs =
+        abi.decode(extraArgs.destChainExtraArgs, (Client.EVMFamilyExtraArgsV1));
+      if (evmExtraArgs.gasLimit > uint256(maxPerMsgGasLimit)) revert MessageGasLimitTooHigh();
```
suggestion: I think we can do some inlining here and remove the `_validateMessage` internal call on L478 as it duplicates the `if (destChainDynamicConfig.familyTag == Client.EVM_FAMILY_TAG)` check.
```solidity
if (destChainDynamicConfig.familyTag == Client.EVM_FAMILY_TAG) {
      Internal._validateEVMAddress(receiver);

      if (abi.decode(extraArgs.destChainExtraArgs, (Client.EVMFamilyExtraArgsV1)).gasLimit > uint256(maxPerMsgGasLimit))
      {
        revert MessageGasLimitTooHigh();
      }
    }
```

elatoskinas replied:
Since the extraArgs have been simplified, I've now moved all the extra args validations to one flow.

The receiver is still validated separately, which I think makes sense, since you could have shared address validations like:

```solidity
if (familyTag == Client.EVM_FAMILY_TAG || familyTag == Client.ZKEVM_FAMILY_TAG) {
  Internal._validateEVMAddress(receiver);
}
```

makramkd commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -267,8 +316,20 @@ library Internal {
+    RampTokenAmount[] tokenAmounts; // array of tokens and amounts to transfer
+  }
+
+  /// @notice Family-agnostic message sent to an OnRamp
```
Small nit: `Family agnostic message sent _from_ an OnRamp` right?

elatoskinas replied:
Yeah `emitted from` would be more accurate

elatoskinas committed:
feat: simplify extra args handling to use EVMExtraArgsV2 directly

RayXpub commented on contracts/src/v0.8/ccip/onRamp/EVM2EVMMultiOnRamp.sol:
```diff
@@ -361,48 +359,97 @@ contract EVM2EVMMultiOnRamp is IEVM2AnyOnRampClient, ITypeAndVersion, OwnerIsCre
+    uint64 nonce = 0;
+    if (
+      destChainConfig.dynamicConfig.familyTag != Client.EVM_FAMILY_TAG
+        || !_parseEVMExtraArgsFromBytes(message.extraArgs, destChainConfig.dynamicConfig).allowOutOfOrderExecution
```
q: what a about the case where we have a non EVM destination chain with an enabled OOO?

elatoskinas replied:
I think we need to flip this - assume OOO by default

But to answer your question - a non-EVM chain can only be configured if it is one of the supported family tags. Currently it is only EVM.

This means that once we onboard a new non-EVM chain, we have to update the contract code, and look at this scenario case-by-case

RayXpub replied:
> But to answer your question - a non-EVM chain can only be configured if it is one of the supported family tags. Currently it is only EVM.

In that case wdyt of enforcing this at the config level and removing it from the hot path for now?

> This means that once we onboard a new non-EVM chain, we have to update the contract code, and look at this scenario case-by-case

👍

elatoskinas replied:
Removed from hot path for now - added `NOTE` where the checks would be required

RayXpub commented on contracts/src/v0.8/ccip/onRamp/EVM2EVMMultiOnRamp.sol:
```diff
@@ -321,21 +322,18 @@ contract EVM2EVMMultiOnRamp is IEVM2AnyOnRampClient, ITypeAndVersion, OwnerIsCre
-    _validateMessage(
-      destChainSelector, message.data.length, extraArgs.gasLimit, numberOfTokens, extraArgs.allowOutOfOrderExecution
-    );
+    // TODO: optimization - consider removing this call, since the Router calls into getFee -> _validateMessage
```
+1

elatoskinas replied:
Will do in next PR

elatoskinas committed:
perf: remove family tag checks for single-condition hot path

elatoskinas commented on contracts/src/v0.8/ccip/libraries/Client.sol:
```diff
@@ -52,4 +52,7 @@ library Client {
   }
+
+  // TODO: pre-optimize by storing the const value
+  bytes4 public constant EVM_FAMILY_TAG = bytes4(keccak256("CCIP EVM Family"));
```
I'll move this to `Internal`

elatoskinas commented on contracts/src/v0.8/ccip/onRamp/EVM2EVMMultiOnRamp.sol:
```diff
@@ -140,7 +141,8 @@ contract EVM2EVMMultiOnRamp is IEVM2AnyOnRampClient, ITypeAndVersion, OwnerIsCre
     uint32 networkFeeUSDCents; //                │ Flat network fee to charge for messages,  multiples of 0.01 USD
-    bool enforceOutOfOrder; // ──────────────────╯ Whether to enforce the allowOutOfOrderExecution extraArg value to be true.
+    bool enforceOutOfOrder; //                   │ Whether to enforce the allowOutOfOrderExecution extraArg value to be true.
+    bytes4 familyTag; // ────────────────────────╯ Tag that identifies the ramp family. Used to determine the correct validations to perform for the family.
```
Could probably rename to `chainFamilySelector`, thoughts?

RayXpub replied:
+1 to be consistent with `destChainSelector`

elatoskinas replied:
Updated

RayXpub commented on contracts/src/v0.8/ccip/onRamp/EVM2EVMMultiOnRamp.sol:
```diff
@@ -361,48 +359,95 @@ contract EVM2EVMMultiOnRamp is IEVM2AnyOnRampClient, ITypeAndVersion, OwnerIsCre
-  function _extraArgsFromBytes(
+  /// @param defaultTxGasLimit default tx gas limit to use in the absence of extra args
+  /// @return EVMExtraArgs the extra args struct (latest version)
+  function _parseUnvalidatedEVMExtraArgsFromBytes(
```
q: can this function be grouped with `_parseEVMExtraArgsFromBytes`? I don't see it being used anywhere else.

elatoskinas replied:
It could, but it's more readable to keep it separate, because there's 3 paths that return `EVMExtraArgs`. We'd have to do something like this:

```solidity
    Client.EVM2ExtraArgsV2 memory extraArgs = Client.EVMExtraArgsV2({
      gasLimit: 0,
      allowOutOfOrderExecution: false
    });

    if (extraArgs.length == 0) {
      // If extra args are empty, generate default values
      extraArgs.gasLimit = defaultTxGasLimit;
    } else {
        bytes4 extraArgsTag = bytes4(extraArgs);
        bytes memory argsData = extraArgs[4:];

      if (extraArgsTag == Client.EVM_EXTRA_ARGS_V2_TAG) {
        extraArgs = abi.decode(...);
      } else if (extraArgsTag == Client.EVM_EXTRA_ARGS_V1_TAG) {
        extraArgs.gasLimit = abi.decode(...);
      }

      revert InvalidExtraArgsTag();
    }

   // ... (remaining logic)
```

imo current approach is more readable and less error-prone (esp. when adding new args in the case when there's a V3)

RayXpub commented on contracts/src/v0.8/ccip/onRamp/EVM2EVMMultiOnRamp.sol:
```diff
@@ -413,24 +458,30 @@ contract EVM2EVMMultiOnRamp is IEVM2AnyOnRampClient, ITypeAndVersion, OwnerIsCre
-    uint256 gasLimit,
     uint256 numberOfTokens,
-    bool allowOutOfOrderExecution
+    bytes memory receiver
```
```suggestion
    bytes calldata receiver
```

RayXpub commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -106,19 +106,30 @@ library Internal {
   /// When abiEncoded, each EVMTokenAmount takes 2 slots, each bytes takes 2 slots, excl bytes contents
   uint256 public constant MESSAGE_FIXED_BYTES_PER_TOKEN = 32 * 4;
 
+  /// @dev AnyEVMMessage struct has 10 fields, including 3 variable arrays.
```
nit: isn't this referring to `Any2EVMRampMessage`?

elatoskinas replied:
Fixed

elatoskinas committed:
refactor: update naming, comments and chain family selector placement

elatoskinas committed:
Merge e8846012710854b36836ad55de159eccd5d7b030 into 54127b1fbc214cf35b150fed03706598267051a6

app-token-issuer-infra-releng[bot] committed:
Update gethwrappers

makramkd committed:
tmp message hasher fix (#1156)

until CCIPMsg changes from chainlink-common are merged in we have to do
this tempfix.

elatoskinas committed:
feat: validate family chain selector on config update

elatoskinas committed:
test: add tests for invalid chain selector family and dest token addresses

elatoskinas committed:
docs: update fixed bytes docs

elatoskinas committed:
chore: re-gen wrappers and snapshots

makramkd commented on contracts/src/v0.8/ccip/onRamp/EVM2EVMMultiOnRamp.sol:
```diff
@@ -361,48 +358,95 @@ contract EVM2EVMMultiOnRamp is IEVM2AnyOnRampClient, ITypeAndVersion, OwnerIsCre
+  /// @param extraArgs The extra args bytes
+  /// @param destChainDynamicConfig Dest chain config to validate against
+  /// @return encodedExtraArgs the parsed & encoded extra args
+  function _convertParsedExtraArgs(
+    bytes calldata extraArgs,
+    DestChainDynamicConfig memory destChainDynamicConfig
+  ) internal pure returns (bytes memory encodedExtraArgs) {
```
q: is this or similar logic going to move to PriceRegistry in future PRs?

elatoskinas replied:
Yeah, planning to move this to Price Reg along with all family-specific configs

makramkd commented on contracts/src/v0.8/ccip/onRamp/EVM2EVMMultiOnRamp.sol:
```diff
@@ -413,24 +457,30 @@ contract EVM2EVMMultiOnRamp is IEVM2AnyOnRampClient, ITypeAndVersion, OwnerIsCre
+  /// @param chainFamilySelector Tag to identify the target family
+  /// @param destAddress Dest address to validate
+  /// @dev precondition - assumes the family tag is correct and validated
+  function _validateDestFamilyAddress(bytes4 chainFamilySelector, bytes memory destAddress) internal pure {
```
Similar q to the above, moving to price reg in future?

elatoskinas replied:
Yup since this is family-specific logic

elatoskinas committed:
Merge branch 'ccip-develop' into feat/evm2any-onramp-structs

elatoskinas committed:
Merge branch 'ccip-develop' into feat/evm2any-onramp-structs
